### PR TITLE
feat(car-recycling): new application

### PR DESCRIPTION
**Pull request by birkirkristmunds**
Car recycling application

## Why
for applicants to apply for car recycling at Recycling fund (Úrvinnslusjóður)

/umsoknir/skilavottord/

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

**Review by brynjarorng — APPROVED** [review]
LGTM

**Review by brynjarorng — CHANGES_REQUESTED** [review]

File: apps/skilavottord/ws/src/app/modules/appSysRecyclingReq/appSysRecyclingReq.resolver.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//@UseGuards(JWTTokenGuard)
+//@Scopes('@urvinnslusjodur.is/skilavottord')
+@Resolver(() => AppSysVehicleInformation)
+export class AppSysRecyclingReqResolver {
+  constructor(private appSysRecyclingReqService: AppSysRecyclingReqService) {}
+
+  //////@Scopes('@island.is/applications/urvinnslusjodur')
+  @Query(() => [AppSysVehicleInformation])
+  async skilavottordAppSysVehicles(): Promise<Array<AppSysVehicleInformation>> {
+    console.log('------------ NEW ENDPOINT HITTTT  1242 ------------------')
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/appSysRecyclingReq/appSysRecyclingReq.resolver.ts]: fixed - Removed

File: apps/skilavottord/ws/src/app/modules/appSysRecyclingReq/appSysRecyclingReq.service.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  async getUserVehiclesInformation(): Promise<AppSysVehicleInformation[]> {
+    return [
+      {
+        permno: 'SR845',
+      },
+      {
+        permno: 'PS888',
+      },
+    ]
+  }
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/appSysRecyclingReq/appSysRecyclingReq.service.ts]: fixed - Removed

File: apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwnerAppSys.resolver.ts [hunk 3]
```diff
@@ -1,23 +1,30 @@
...
   @Mutation(() => Boolean)
   async createSkilavottordVehicleOwnerAppSys(
     @CurrentUser() user: User,
-    @Args('name') name: string,
-  ) {
+    @Args('input') input: CreateOwnerInput,
+  ): Promise<boolean> {
+    console.log('createSkilavottordVehicleOwnerAppSys', input.name)
+    console.log('user', user)
+    console.log('NATIONAL ID', user.nationalId.replace('**REMOVE_PII:', ''))
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwnerAppSys.resolver.ts]: fixed - Removed

File: apps/skilavottord/web/next.config.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ const withVanillaExtract = createVanillaExtractPlugin()
 
 const {
   BASE_PATH = '/app/skilavottord',
-  API_URL = 'http://localhost:3333',
+  API_URL = 'http://localhost:3339',
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/web/next.config.js:7]: The reason for this was to be able run both skilavottord-ws server and Application system on the same machine locally. Do you recommend that we revert it back to 3333 before pushing on main?

File: apps/skilavottord/ws/src/app/modules/vehicle/vehicleAppSys.resolver.ts [hunk 5]
```diff
@@ -81,6 +88,8 @@ export class VehicleAppSysResolver {
     @CurrentUser() user: User,
     @Args('permno') permno: string,
   ) {
+    console.log('createSkilavottordVehicleAppSys')
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/vehicle/vehicleAppSys.resolver.ts]: Fixed - Removed

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.module.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    //   provide: RecyclingFundClientService,
+    //   //useFactory: () => {
+    //   useFactory: (idsClientConfig: ConfigType<typeof IdsClientConfig>) => {
+    //     const api = new RecyclingFundGraphQLClientApi(
+    //       new Configuration(configFactory(idsClientConfig)),
+    //     )
+    //
+    //     return new RecyclingFundClientService(api)
+    //   },
+    // },
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/clients/recycling-fund/src/lib/recyclingFundClient.module.ts]: fixed - Removed

File: apps/skilavottord/ws/src/environments/environment.ts [hunk 7]
```diff
@@ -2,8 +2,9 @@ const devConfig = {
   production: false,
   auth: {
     issuer: 'https://identity-server.dev01.devland.is',
-    audience: '@urvinnslusjodur.is/skilavottord',
     nextAuthCookieName: 'next-auth.session-token',
+    audience: '@urvinnslusjodur.is/skilavottord',
+    // audience: ['@urvinnslusjodur.is/skilavottord', '@island.is'],
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/environments/environment.ts]: fixed - Removed

File: apps/skilavottord/web/next.config.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/web/next.config.js]: We rollbacked the port to 3333

File: apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts [hunk 8]
```diff
@@ -15,6 +15,10 @@ export class VehicleOwnerResolver {
     @CurrentUser() user: User,
     @Args('name') name: string,
   ) {
+    console.log('!!!!!!!! OLD createSkilavottordVehicleOwner   !!!!!!!!')
+    console.log('User', user)
+    console.log('name', name)
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts]: Fixed - removed

File: apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts [hunk 9]
```diff
@@ -5,7 +5,7 @@ import { Authorize, CurrentUser, User } from '../auth'
 import { VehicleOwnerModel } from './vehicleOwner.model'
 import { VehicleOwnerService } from './vehicleOwner.service'
 
-@Authorize()
+//@Authorize()
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts]: Fixed removed

File: apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts [hunk 10]
```diff
@@ -1,11 +1,10 @@
 import { Args, Resolver, Mutation } from '@nestjs/graphql'
 
-import { Authorize, CurrentUser, User } from '../auth'
+import { CurrentUser, User } from '../auth'
 
 import { VehicleOwnerModel } from './vehicleOwner.model'
 import { VehicleOwnerService } from './vehicleOwner.service'
 
-@Authorize()
 @Resolver(() => VehicleOwnerModel)
```
Comment (essgje) [inline apps/skilavottord/ws/src/app/modules/vehicleOwner/vehicleOwner.resolver.ts:8]: Ath @Authorize()

**Review by AndesKrrrrrrrrrrr — APPROVED** [review]
E2E files LGTM

File: apps/api/src/app/app.module.ts [hunk 11]
```diff
@@ -86,6 +86,7 @@ import { DisabilityLicenseClientConfig } from '@island.is/clients/disability-lic
 import { DrivingLicenseApiConfig } from '@island.is/clients/driving-license'
 import { DrivingLicenseBookClientConfig } from '@island.is/clients/driving-license-book'
 import { ElectronicRegistrationsClientConfig } from '@island.is/clients/electronic-registration-statistics'
+
```
Comment (saevarma) [inline apps/api/src/app/app.module.ts:89]: Unnecessary change which touches the `api` project which otherwise wouldn't be affected.

File: apps/skilavottord/ws/src/app/app.module.ts [hunk 12]
```diff
@@ -38,6 +39,10 @@ const autoSchemaFile = environment.production
       useClass: SequelizeConfigService,
     }),
     AuthModule,
+    AuthJwtModule.register({
+      audience: null,
+      issuer: 'https://identity-server.dev01.devland.is',
```
Comment (saevarma) [inline apps/skilavottord/ws/src/app/app.module.ts:44]: This config needs to be environment based.

File: apps/skilavottord/ws/src/app/modules/samgongustofa/samgongustofa.resolver.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import { Query, Resolver } from '@nestjs/graphql'
+import { Query, Resolver, Args } from '@nestjs/graphql'
```
Comment (saevarma) [inline apps/skilavottord/ws/src/app/modules/samgongustofa/samgongustofa.resolver.ts:1]: Looks like a redundant change.

File: libs/auth-nest-tools/src/lib/ids-user.guard.ts
Comment (saevarma) [inline libs/auth-nest-tools/src/lib/ids-user.guard.ts:1]: Redundant changes.

File: libs/auth/scopes/src/lib/recycling-fund.scope.ts [hunk 14]
```diff
@@ -0,0 +1,3 @@
+export enum RecyclingFundScope {
+  recyclingFund = '@urvinnslusjodur.is/skilavottord',
```
Comment (saevarma) [inline libs/auth/scopes/src/lib/recycling-fund.scope.ts:2]: The organisation is `RecyclingFund` but this specific scope is for car recycling (skilavottord)
```suggestion
  carRecycling = '@urvinnslusjodur.is/skilavottord',
```

File: libs/auth/scopes/src/lib/api.scope.ts [hunk 15]
```diff
@@ -18,4 +18,5 @@ export enum ApiScope {
   health = '@island.is/health',
   workMachines = '@island.is/work-machines',
   samgongustofaVehicles = '@island.is/applications/samgongustofa-vehicles',
+  recyclingFund = '@island.is/applications/urvinnslusjodur',
```
Comment (saevarma) [inline libs/auth/scopes/src/lib/api.scope.ts:21]:
```suggestion
  carRecycling = '@island.is/applications/urvinnslusjodur',
```

File: libs/clients/recycling-fund/.eslintrc.json
Comment (saevarma) [inline libs/clients/recycling-fund/.eslintrc.json:1]: This library should be named `car-recycling` as it is a client library to communicate with car recycling (skilavottord)  gql. RecyclingFund the organisation could have more apis which  we would integrate with later on in another client library.

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+          input: {
+            permno: 'THS45',
+            requestType: RecyclingRequestTypes.pendingRecycle,
+          },
+        } as CreateRequestMutationVariables,
+      }),
+    })
+
+    if (!response.ok) {
+      throw new Error(`Failed to creating recycling request from`)
```
Comment (saevarma) [inline libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts:65]: This looks like an incomplete error message

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    private readonly fetch: EnhancedFetchAPI,
+  ) {}
+
+  async createRecyclingFundRequest(user: User, body: any) {
+    return await this.fetch(this.config.gqlBasePath, {
+      ...baseGqlRequestOptions,
+      auth: user,
+      body: JSON.stringify(body),
+    })
+  }
```
Comment (saevarma) [inline libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts:46]: The name should be more generic like `gqlRequestWithAuth` and `body` shouldn't be an `any` type. You should be able to type it and use generics to help with typesafety of `query` and `variabes` fields.

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    if (!response.ok) {
+      throw new Error(`Failed to creating recycling request from`)
+    }
+
+    const data = (await response.json()) as CreateRequestMutation
+
+    console.log('response from createRecyclingRequest GQL', {
+      status: response.status,
+      data,
+    })
```
Comment (saevarma) [inline libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts:73]: clean up console.log?

File: libs/clients/recycling-fund/src/lib/utils.ts
Comment (saevarma) [inline libs/clients/recycling-fund/src/lib/utils.ts:1]: We can delete this?

File: libs/clients/recycling-fund/src/lib/recyclingFundFetchProvider.ts [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+  createEnhancedFetch,
+  type EnhancedFetchAPI,
+} from '@island.is/clients/middlewares'
+import {
+  type ConfigType,
+  IdsClientConfig,
+  LazyDuringDevScope,
+} from '@island.is/nest/config'
+
+export const RecyclingFundFetchKey = 'RecyclingFundFetchProviderKey'
```
Comment (saevarma) [inline libs/clients/recycling-fund/src/lib/recyclingFundFetchProvider.ts:14]:
```suggestion
export const RecyclingFundFetchProviderKey = 'RecyclingFundFetchProviderKey'
```

**Review by saevarma — COMMENTED** [review]
I skipped reviewing application system specific files.

In the new `*AppSys.resolver.ts` in skilavottord-ws are all of the methods actually used? Feels like there are more than the application is actually using.

File: apps/skilavottord/ws/src/app/modules/samgongustofa/samgongustofa.resolver.ts (hunk 13 above)
Comment (essgje) [reply apps/skilavottord/ws/src/app/modules/samgongustofa/samgongustofa.resolver.ts]: Removed

**Review by thordurhhh — APPROVED** [review]
Code owned files LGTM.

File: apps/skilavottord/ws/src/app/app.module.ts (hunk 12 above)
Comment (birkirkristmunds) [reply apps/skilavottord/ws/src/app/app.module.ts]: fixed

File: libs/auth-nest-tools/src/lib/ids-user.guard.ts
Comment (birkirkristmunds) [reply libs/auth-nest-tools/src/lib/ids-user.guard.ts:1]: code changes rollbacked

File: libs/auth/scopes/src/lib/recycling-fund.scope.ts (hunk 14 above)
Comment (birkirkristmunds) [reply libs/auth/scopes/src/lib/recycling-fund.scope.ts]: Fixes

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts (hunk 16 above)
Comment (birkirkristmunds) [reply libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts]: Code removed

File: libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts (hunk 18 above)
Comment (birkirkristmunds) [reply libs/clients/recycling-fund/src/lib/recyclingFundClient.service.ts]: Code removed

File: libs/auth/scopes/src/lib/api.scope.ts (hunk 15 above)
Comment (birkirkristmunds) [reply libs/auth/scopes/src/lib/api.scope.ts]: fixed

File: libs/clients/recycling-fund/.eslintrc.json
Comment (birkirkristmunds) [reply libs/clients/recycling-fund/.eslintrc.json:1]: Renamed the client library to car-recycling

File: libs/clients/recycling-fund/src/lib/utils.ts
Comment (birkirkristmunds) [reply libs/clients/recycling-fund/src/lib/utils.ts:1]: File deleted

File: libs/application/templates/driving-license/src/forms/prerequisites/sectionExternalData.ts [hunk 20]
```diff
@@ -25,48 +25,7 @@ export const sectionExternalData = buildSubSection({
       id: 'approveExternalData',
       subTitle: m.externalDataSubTitle,
       checkboxLabel: m.externalDataAgreement,
-      dataProviders: [
```
Comment (HelgiHH) [inline libs/application/templates/driving-license/src/forms/prerequisites/sectionExternalData.ts:28]: this seems a bit off, do you need to remove sections from another application?

Comment (birkirkristmunds) [reply libs/application/templates/driving-license/src/forms/prerequisites/sectionExternalData.ts]: absolutly not, This was a mistake 
Thanks for catching this!

**Review by brynjarorng — APPROVED** [review]
LGTM

File: libs/auth-nest-tools/src/lib/ids-user.guard.ts
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/auth-nest-tools/src/lib/ids-user.guard.ts:1]: Rollbacked redundant changes

**Review by HelgiHH — APPROVED** [review]

File: libs/clients/car-recycling/README.md
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/clients/car-recycling/README.md:1]: Readme updated

File: libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+        this.recycleVehicles(auth, vehicle, RecyclingRequestTypes.cancelled)
+      })
+    } catch (error) {
+      isError = true
+      this.logger.error(error.messages)
+    }
+
+    return new Promise((resolve, reject) => {
+      // Simulate an asynchronous operation (e.g., fetching data)
+      setTimeout(() => {
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts:120]: Why the extra timeout ?

File: libs/application/templates/car-recycling/src/dataProviders/index.ts [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  NationalRegistrySpouseApi,
+} from '@island.is/application/types'
+
+interface CurrentVehiclesParameters {
+  showOwned?: boolean
+  showCoOwned?: boolean
+  showOperated?: boolean
+}
+
+export const CurrentVehiclesApi = defineTemplateApi<CurrentVehiclesParameters>({
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/application/templates/car-recycling/src/dataProviders/index.ts]: Fixed, removed it and all code related to it.

File: libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts (hunk 21 above)
Comment (birkirkristmunds) [reply libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts:120]: Im trying to prevent the application to be sent if there is an error coming from the backend.  Since we are making more than one call to the backend when sending the application I don't have a promise to return. 
It seems that if I removed the setTimeout then the application is submitted and no error is shown.

File: libs/application/templates/car-recycling/src/lib/CarRecyclingTemplate.spec.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    status: ApplicationStatus.IN_PROGRESS,
+    modified: new Date(),
+    applicantActors: [],
+    answers,
+    state,
+    externalData,
+  }
+}
+
+describe('Old Age Pension Template', () => {
```
[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/application/templates/car-recycling/src/lib/CarRecyclingTemplate.spec.ts:33]: Fixed

File: libs/application/templates/car-recycling/src/lib/answerValidationSections/vehicles.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
+import { Application } from '@island.is/application/types'
+
+import { AnswerValidationConstants } from '../../shared/constants'
+import { buildError } from './utils'
+import { VehicleMiniDto } from '@island.is/clients/vehicles'
+import { errorMessages } from './../messages'
+
+export const vehicles = (newAnswer: unknown, application: Application) => {
```
[the remark this answers is not in the document]
Comment (Thorrik58) [reply libs/application/templates/car-recycling/src/lib/answerValidationSections/vehicles.ts:8]: I would encourage you to use zod validation for this case. If it doesnt seem feasible I would like to give it a try as we believe currently that most cases should be possible to tackle with zod. It would take me a day or two to free up my backlog tho. We are trying to phase out answer validators, but it might not be possible.

[the remark this answers is not in the document]
Comment (birkirkristmunds) [reply libs/application/templates/car-recycling/src/lib/answerValidationSections/vehicles.ts]: We have found the solution for this and removed the answer validators, so it is part of this PR

File: libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts (hunk 21 above)
Comment (obmagnusson) [reply libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts:120]: If waiting for all operations to complete is the goal, it should be tied to the actual completion of the operations, not an arbitrary timeout. Suggest just using a promise.all to properly await all executions and resolve accordingly.

Comment (birkirkristmunds) [reply libs/application/template-api-modules/src/lib/modules/templates/car-recycling/car-recycling.service.ts]: I have made changes to the code as you suggested and was able to remove the timeout
